Rework JDT completions via LSP

BoykoAlex opened the pull request:
Rework naive implementation of LSP completions for Java Editor

martinlippert reviewed (approved):
Tried the changes with our Spring Boot language server in the various situations where we use Spring completions inside of Java source code, seems to work nicely.

rubenporras commented on org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/completion/LSCompletionProposal.java:
```diff
@@ -306,10 +323,19 @@ public String getDisplayString() {
 	public StyledString getStyledDisplayString() {
-		if (Boolean.TRUE.equals(item.getDeprecated())) {
-			return new StyledString(getDisplayString(), StyleUtil.DEPRECATE);
+		StyledString ss = Boolean.TRUE.equals(item.getDeprecated())
```
would you mind doing this part of the PR (which I think adds support for completion label details) in a separate PR? That part should be very easy to review then.

BoykoAlex replied:
Agreed.

rubenporras commented on org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/completion/LSCompletionProposal.java:
```diff
@@ -122,7 +123,7 @@ public class LSCompletionProposal
 	protected int currentOffset = -1;
 	protected @Nullable ITextViewer viewer;
-	private final IDocument document;
+	protected final IDocument document;
```
I am not following this part of the PR. Why does it need to be protected?

BoykoAlex replied:
Indeed... was a rogue change... :-\

rubenporras commented on org.eclipse.lsp4e/META-INF/MANIFEST.MF:
```diff
@@ -2,7 +2,7 @@ Manifest-Version: 1.0
 Bundle-Name: Language Server Protocol client for Eclipse IDE (Incubation)
 Bundle-SymbolicName: org.eclipse.lsp4e;singleton:=true
-Bundle-Version: 0.18.20.qualifier
+Bundle-Version: 0.19.0.qualifier
```
could not just be 0.18.21 following what we did so far?

rubenporras commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaProposal.java:
```diff
@@ -94,18 +84,97 @@ public int getRelevance() {
+
+		if (infoControlCreator == null) {
+			/*
+			 * FIXME: Take control creators (and link handling) out of JavadocHover,
```
This has been marked as RESOLVED WONTFIX, so probably it will never happen. Should we remove the FIXME and leave the link to the bug report?

BoykoAlex replied:
Remove part of the comment

rubenporras commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaProposal.java:
```diff
@@ -11,18 +11,36 @@
+import org.eclipse.core.runtime.Platform;
 import org.eclipse.jdt.annotation.Nullable;
 import org.eclipse.jdt.internal.codeassist.RelevanceConstants;
+import org.eclipse.jdt.internal.ui.JavaPlugin;
```
@mickaelistria , should we be avoiding importing internal classes from the JDT or is this something that is not an issue if you are part of the SimRel?

mickaelistria replied:
Yes, better avoiding it (when possible)

mickaelistria commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaProposal.java:
```diff
@@ -94,18 +84,97 @@ public int getRelevance() {
+	 */
+	@Override
+	public @Nullable IInformationControlCreator getInformationControlCreator() {
+		Shell shell= JavaPlugin.getActiveWorkbenchShell();
```
That part can probably be inlined instead of referencing the internal.

BoykoAlex replied:
Done

mickaelistria commented on org.eclipse.lsp4e.jdt/src/org/eclipse/lsp4e/jdt/LSJavaProposal.java:
```diff
@@ -94,18 +84,97 @@ public int getRelevance() {
+	 */
+	private @Nullable String getCSSStyles() {
+		if (fgCSSStyles == null) {
+			Bundle bundle= Platform.getBundle(JavaPlugin.getPluginId());
```
Would better use `FrameworkUtil.getBundle(IJavaEditorTextHover.class)` (an API class related to the file we want)

BoykoAlex commented:
@rubenporras I have separated out label details, addressed your comments (Thank you very much for the comments by the way). Also cleaned up a few things. Lets see what else stands out to you :-)

BoykoAlex commented on org.eclipse.lsp4e/src/org/eclipse/lsp4e/operations/completion/LSCompletionProposal.java:
```diff
@@ -306,10 +323,19 @@ public String getDisplayString() {
 	public StyledString getStyledDisplayString() {
-		if (Boolean.TRUE.equals(item.getDeprecated())) {
-			return new StyledString(getDisplayString(), StyleUtil.DEPRECATE);
+		StyledString ss = Boolean.TRUE.equals(item.getDeprecated())
```
Yes, sounds good
